src/rate_limiter.rs

```diff
@@ -181,89 +277,276 @@ impl RateLimiter {
+    ) -> Option<u64> {
+
+        if let MessageContent::UserMessagePart {
+            ref msg_id,
+            part_count,
+            part_index,
+            ref payload,
+            ..
+        } = *msg
+        {
+            rate_limiter.deduct_response(client, msg_id, part_count, part_index, payload)
+        } else {
+            panic!("message is not a UserMessagePart: {:?}", msg);
         }
     }
 
-    /// Checks that a single client cannot exceed its individual limit and that its throughput is
+    // Generate a single `UserMessagePart` for a random `GetIData` request.
+    fn random_payload() -> MessageContent {
+        let user_message = UserMessage::Request(Request::GetIData {
+            name: rand::random(),
+            msg_id: MessageId::new(),
+        });
+        let message_parts = unwrap!(user_message.to_parts(0));
+        assert_eq!(message_parts.len(), 1);
+        message_parts[0].clone()
+    }
+
+    /// Checks that a single client cannot exceed the proxy's soft limit and that its throughput is
     /// the full rate of the rate-limiter.
     #[test]
     fn single_client() {
         let mut rate_limiter = RateLimiter::new(false);
         let client = IpAddr::from([0, 0, 0, 0]);
 
         // Consume full allowance.
-        assert_large_message_can_be_added(&mut rate_limiter, &client);
+        assert!(huge_message_can_be_added(&mut rate_limiter, &client));
 
         // Check client can't add any more requests just now.
-        assert_small_message_cannot_be_added(&mut rate_limiter, &client);
+        assert!(small_message_cannot_be_added(&mut rate_limiter, &client));
 
         // Advance the clock 1ms and check the small request can now be added, but the large request
         // is still disallowed.
         FakeClock::advance_time(1);
-        assert_small_message_can_be_added(&mut rate_limiter, &client);
-        assert_large_message_cannot_be_added(&mut rate_limiter, &client);
+        assert!(small_message_can_be_added(&mut rate_limiter, &client));
+        assert!(huge_message_cannot_be_added(&mut rate_limiter, &client));
 
         // Advance the clock enough to allow the client's entry to fully drain away. (No need to
         // round up the calculation here as we've already advanced by 1ms which is equivalent to
         // rounding up the millisecond calculation).
-        let wait_millis = CLIENT_CAPACITY * 1000 / RATE as u64;
+        let wait_millis = SOFT_CAPACITY * 1000 / RATE as u64;
         FakeClock::advance_time(wait_millis);
-        assert_large_message_can_be_added(&mut rate_limiter, &client);
+        assert!(huge_message_can_be_added(&mut rate_limiter, &client));
+    }
+
+    #[test]
+    fn overcharge_correction() {
+        let mut rate_limiter = RateLimiter::new(false);
+        let client = IpAddr::from([0, 0, 0, 0]);
+
+        let approx_num_parts = 10;
+        let data_size = approx_num_parts * MAX_PART_LEN;
+        let data = ImmutableData::new(vec![0; data_size]);
```

Comment:
better not to create an `exact sized` data, instead something like : 
```rust
let data_size = rng.gen();
let num_parts = data_size / MAX_PART_LEN```
If we can't refund accurately, it shall be reflected in the testing code.